File: src/custom/spiral/spiral.c
/*
 * spiral.c -- Spiral: a delay whose repeats climb (or sink), MS-70CDR.
 *
 * Pedal port of tools/audio_preview/renderers/spiral.py. A grain-re-triggered
 * tap is read at `ratio` INSIDE the feedback loop, so one reader both delays
 * and pitch-shifts (Hydra's trick) and the shift COMPOUNDS: repeat 1 is one
 * step away, repeat 2 two steps, an endless staircase fading with the feedback.
 *
 * Two controls sit on that ratio, covering both readings of "the delay rises":
 *   Rise  a constant ratio != 1 -> discrete per-repeat staircase. Bipolar,
 *         unity at centre; left of centre the repeats sink instead.
 *   Glide+Span  an onset-triggered ramp on the same ratio. A note resets the
 *         ramp to ZERO and it climbs toward Glide's depth over Span, so the
 *         repeats begin at pitch and drift away over a while -- a riser, NOT a
 *         swoop that settles back.
 * Glide 0 = pure staircase; Rise centred + Glide up = pure glide, no net
 * transposition; both = a climb that also swoops.
 *
 * 8 knobs. Div and Tempo sit on knobs 1-2 deliberately: those are the two
 * verbatim-stock LineSel edit handlers, the most reliable pair in the pack, and
 * they are the two you actually dial when playing to a drum machine.
 *   Div    (knob 1) musical division of Tempo: 1/32, 1/16T, 1/16, 1/8T, 1/8,
 *                   dotted 1/8, 1/4, dotted 1/4
 *   Tempo  (knob 2) BPM, 40..240, read straight off the knob so the pedal's own
 *                   number IS the tempo
 *   Feedbk (knob 3) repeats, scaled to 0.965 and soft-clipped in the loop
 *   Rise   (knob 4) per-repeat ratio: 0 = oct down, 50 = unity, 100 = oct up
 *   Glide  (knob 5) depth of the onset ramp on that ratio
 *   Span   (knob 6) how long the rise takes, ~0.19s .. ~8s
 *   Tone   (knob 7) low-pass inside the loop
 *   Mix    (knob 8) dry/wet
 *
 * Why tempo and not a free delay time: the old Time knob was a squared map onto
 * 93ms..1.2s, so every tempo change meant re-dialling by ear. Time=17 at 120 BPM
 * measured 124.9 ms -- a 1/16 note to within 0.1 ms -- so that setting is now
 * simply Div=1/16, and it tracks when the drum machine's tempo moves.
 *
 * Safe-DSP: no math lib -- the ratio map is piecewise LINEAR rather than
 * 2^(semis/12), and Span's one-pole coefficient is a CUBIC in the knob instead
 * of 1-exp(-1/(t*sr)), because both exp2 and the divide would pull in helper
 * calls. No runtime divide (grain length is a fixed power of two so 1/G is a
 * literal), no modulo (power-of-two ring mask), no switch (jump tables are
 * unreachable in a ZDL and hard-freeze the DSP), no static arrays, no
 * float->unsigned casts. Ring lives in the ctx[3] arena, cleared once at init.
 * ctx[11]/ctx[12] magic shuttle preserved.
 *
 * Tap-overtake guard: the tap starts D behind the write head and advances at
 * `ratio`, so over one grain it gains G*(ratio-1) on the writer. With G=1024,
 * ratio clamped to 4.0 and D >= 4096, the worst case gain is 3072 < 4096 -- the
 * tap can never read past the write head into unwritten samples.
 */

#include <stdint.h>

#include "../../airwindows/common/zoom_params.h"
#include "spiral_params.h"

#ifndef SPIRAL_AUDIO_FUNC
#define SPIRAL_AUDIO_FUNC Fx_DLY_Spiral
#endif

#define SPIRAL_DO_PRAGMA(x) _Pragma(#x)
#define SPIRAL_EXPAND_PRAGMA(x) SPIRAL_DO_PRAGMA(x)
#define SPIRAL_CODE_SECTION(f) SPIRAL_EXPAND_PRAGMA(CODE_SECTION(f, ".audio"))

SPIRAL_CODE_SECTION(SPIRAL_AUDIO_FUNC)

#define ZDL_PTR(type, word) ((type)(uintptr_t)(word))

#define SP_MAGIC   0x53505241u   /* 'SPRA' */
#define SP_VERSION 1u

#define SP_BUF       131072u     /* 512 KB, ~2.97 s */
#define SP_MASK      (SP_BUF - 1u)
#define SP_D_FLOOR   1536        /* ~35 ms. Was 4096: a FIXED tap-overtake floor that
                                  * made 1/16 impossible above ~161 BPM and 1/32
                                  * impossible at any tempo. The guard is now derived
                                  * from D per block (see ratioMax below), so the floor
                                  * only has to keep ratioMax comfortably above 1. */
#define SP_D_CEIL    120000      /* ~2.72 s, inside the 2.97 s ring with headroom */
#define SP_SR_X60    2646000.0f  /* 44100 * 60: samples per quarter = this / BPM */
#define SP_BPM_MIN   40.0f
#define SP_BPM_MAX   240.0f
/* Quadratic seed for 1/BPM through 40/140/240, refined by 4 Newton-Raphson
 * steps (r = r*(2 - B*r)). Multiplies only -- a runtime divide would pull in
 * __c6xabi_divf, which is a freeze class. Worst relative error over the whole
 * range is 2.0e-9, i.e. exact for audio. */
#define SP_RSEED_A   0.03630952f
#define SP_RSEED_B   (-3.125e-4f)
#define SP_RSEED_C   7.440476e-7f
#define SP_GRAIN     1024        /* fixed power of two ... */
#define SP_INV_GRAIN 9.765625e-4f /* ... so 1/G is a literal, not a divide */
#define SP_RATIO_MIN 0.25f
#define SP_RATIO_MAX 4.0f
#define SP_ENV_FAST  0.02f
#define SP_ENV_SLOW  0.0008f
#define SP_ONSET_MUL 1.7f
#define SP_ONSET_ADD 0.008f
#define SP_COOLDOWN  2205        /* ~50 ms between onset triggers */
#define SP_K_MAX     1.13e-4f    /* Span short end (~0.19 s) */
#define SP_K_MIN     2.8e-6f     /* Span long end  (~8.1 s) */
#define SP_DENORM    1.0e-18f

typedef struct SpiralState {
    uint32_t magic;
    uint32_t version;
    uint32_t initialized;
    uint32_t writePos;
    float tap;
    int32_t grainT;
    float lp;
    float envFast, envSlow, ramp;
    int32_t cool;
} SpiralState;

static inline float sp_soft(float x)
{
    if (x > 1.0f) return 1.0f;
    if (x < -1.0f) return -1.0f;
    return 1.5f * x - 0.5f * x * x * x;
}

static inline float sp_abs(float x) { return x < 0.0f ? -x : x; }

/* Denormals are a documented freeze/stall class on this DSP and only `lp` was
 * ever flushed. `ramp` is the worst offender: it converges as
 * ramp += k*(1-ramp), so once it settles, (1 - ramp) is a denormal on EVERY
 * sample thereafter -- which is exactly why the rise "stopped working after it
 * had been on for a while". The envelopes do the same during silence. */
static inline float sp_flush(float x)
{
    if (x < SP_DENORM && x > -SP_DENORM) return 0.0f;
    return x;
}

static inline float sp_read(const float *buf, float pos)
{
    int32_t i0 = (int32_t)pos;
    float fr = pos - (float)i0;
    uint32_t a = ((uint32_t)i0) & SP_MASK;
    uint32_t b = (a + 1u) & SP_MASK;
    return buf[a] * (1.0f - fr) + buf[b] * fr;
}

static inline float sp_wrap(float pos)
{
    if (pos >= (float)SP_BUF) pos -= (float)SP_BUF;
    if (pos < 0.0f) pos += (float)SP_BUF;
    return pos;
}

void SPIRAL_AUDIO_FUNC(unsigned int *ctx)
{
    float *params = ZDL_PTR(float *, ctx[1]);
    float *fxBuf  = ZDL_PTR(float *, ctx[5]);

    unsigned int *magicSrc = ZDL_PTR(unsigned int *, ctx[12]);
    unsigned int *magicDst = ZDL_PTR(unsigned int *, *(unsigned int *)ZDL_PTR(unsigned int *, ctx[11]));
    *magicDst = *magicSrc;

    if (params[0] < 0.5f) return;

    volatile unsigned int *desc = ZDL_PTR(volatile unsigned int *, ctx[3]);
    if (!desc) return;
    uintptr_t base = (uintptr_t)desc[0];
    uintptr_t end  = (uintptr_t)desc[1];
    if (base == 0u || end <= base) return;
    if ((base & 3u) != 0u) return;

    uintptr_t bufBase = (base + sizeof(SpiralState) + 7u) & ~(uintptr_t)7u;
    if (bufBase + (uintptr_t)SP_BUF * 4u > end) return;

    SpiralState *st = (SpiralState *)base;
    float *buf = (float *)bufBase;

    if (st->magic != SP_MAGIC || st->version != SP_VERSION || !st->initialized) {
        st->magic = SP_MAGIC;
        st->version = SP_VERSION;
        st->writePos = 0u;
        st->tap = 0.0f;
        st->grainT = 0;
        st->lp = 0.0f;
        st->envFast = st->envSlow = 0.0f;
        st->ramp = 1.0f;
        st->cool = 0;
        uint32_t i;
        for (i = 0u; i < SP_BUF; i++) buf[i] = 0.0f;
        st->initialized = 1u;
    }

    float dv = zoom_param_norm01(params[SPIRAL_DIV_SLOT],    SPIRAL_DIV_DEFAULT_NORM);
    float tp = zoom_param_norm01(params[SPIRAL_TEMPO_SLOT],  SPIRAL_TEMPO_DEFAULT_NORM);
    float fn = zoom_param_norm01(params[SPIRAL_FEEDBK_SLOT], SPIRAL_FEEDBK_DEFAULT_NORM);
    float rn = zoom_param_norm01(params[SPIRAL_RISE_SLOT],   SPIRAL_RISE_DEFAULT_NORM);
    float gn = zoom_param_norm01(params[SPIRAL_GLIDE_SLOT],  SPIRAL_GLIDE_DEFAULT_NORM);
    float sn = zoom_param_norm01(params[SPIRAL_SPAN_SLOT],   SPIRAL_SPAN_DEFAULT_NORM);
    float on = zoom_param_norm01(params[SPIRAL_TONE_SLOT],   SPIRAL_TONE_DEFAULT_NORM);
    float mix = zoom_param_norm01(params[SPIRAL_MIX_SLOT],   SPIRAL_MIX_DEFAULT_NORM);

    /* Tempo knob reads straight as BPM so the pedal's own number is the tempo. */
    float bpm = tp * SP_BPM_MAX;
    if (bpm < SP_BPM_MIN) bpm = SP_BPM_MIN;
    else if (bpm > SP_BPM_MAX) bpm = SP_BPM_MAX;

    float r = SP_RSEED_A + SP_RSEED_B * bpm + SP_RSEED_C * bpm * bpm;
    r = r * (2.0f - bpm * r);
    r = r * (2.0f - bpm * r);
    r = r * (2.0f - bpm * r);
    r = r * (2.0f - bpm * r);
    float quarter = SP_SR_X60 * r;          /* samples per quarter note */

    /* Division ladder. if/else, never switch: a jump table is unreachable in a
     * ZDL and hard-freezes the DSP. */
    float divMul;
    if      (dv < 0.125f) divMul = 0.125f;                  /* 1/32          */
    else if (dv < 0.250f) divMul = 0.1666667f;              /* 1/16 triplet  */
    else if (dv < 0.375f) divMul = 0.25f;                   /* 1/16          */
    else if (dv < 0.500f) divMul = 0.3333333f;              /* 1/8 triplet   */
    else if (dv < 0.625f) divMul = 0.5f;                    /* 1/8           */
    else if (dv < 0.750f) divMul = 0.75f;                   /* dotted 1/8    */
    else if (dv < 0.875f) divMul = 1.0f;                    /* 1/4           */
    else                  divMul = 1.5f;                    /* dotted 1/4    */

    int32_t D = (int32_t)(quarter * divMul);
    if (D < SP_D_FLOOR) D = SP_D_FLOOR;
    else if (D > SP_D_CEIL) D = SP_D_CEIL;

    /* Tap-overtake guard, now derived rather than assumed. The tap starts D
     * behind the writer and gains G*(ratio-1) over one grain, so it stays behind
     * as long as ratio <= 1 + (D - G)/G. Deriving it per block is what lets the
     * short divisions exist at all. */
    float ratioMax = 1.0f + (float)(D - SP_GRAIN) * SP_INV_GRAIN;
    if (ratioMax > SP_RATIO_MAX) ratioMax = SP_RATIO_MAX;
    else if (ratioMax < 1.05f)   ratioMax = 1.05f;
    float fb = fn * 0.965f;

    /* piecewise linear, unity at centre -- no exp2 */
    float baseRatio;
    if (rn < 0.5f) baseRatio = 0.5f + rn;
    else           baseRatio = 1.0f + (rn - 0.5f) * 2.0f;
    float glideDir = (rn >= 0.5f) ? 1.0f : -1.0f;
    float glideAmt = gn * 0.6f;

    /* Span -> one-pole coefficient. Cubic in the knob so the long end gets the
     * resolution, and it avoids both exp() and a divide. */
    float inv = 1.0f - sn;
    float rampK = SP_K_MIN + SP_K_MAX * inv * inv * inv;

    float lpCoef = 0.03f + on * 0.5f;

    uint32_t wp  = st->writePos;
    float tap    = st->tap;
    int32_t gt   = st->grainT;
    float lp     = st->lp;
    float envF   = st->envFast;
    float envS   = st->envSlow;
    float ramp   = st->ramp;
    int32_t cool = st->cool;

    int i;
    for (i = 0; i < 8; i++) {
        float dry = 0.5f * (fxBuf[i] + fxBuf[i + 8]);

        float ax = sp_abs(dry);
        envF += SP_ENV_FAST * (ax - envF);
        envS += SP_ENV_SLOW * (ax - envS);
        if (cool <= 0 && envF > envS * SP_ONSET_MUL + SP_ONSET_ADD) {
            ramp = 0.0f;                 /* a note restarts the climb from pitch */
            cool = SP_COOLDOWN;
        }
        if (cool > 0) cool--;
        ramp += rampK * (1.0f - ramp);   /* ... and it rises from there over Span */

        float ratio = baseRatio * (1.0f + glideAmt * ramp * glideDir);
        if (ratio < SP_RATIO_MIN) ratio = SP_RATIO_MIN;
        else if (ratio > ratioMax) ratio = ratioMax;

        if (gt >= SP_GRAIN) {            /* re-trigger the tap D behind */
            gt = 0;
            tap = sp_wrap((float)(int32_t)wp - (float)D);
        }

        /* triangular grain window, smoothstepped */
        float e;
        if (gt < (SP_GRAIN >> 1)) e = (float)gt * (2.0f * SP_INV_GRAIN);
        else                      e = (float)(SP_GRAIN - gt) * (2.0f * SP_INV_GRAIN);
        if (e > 1.0f) e = 1.0f;
        e = e * e * (3.0f - 2.0f * e);

        float echo = sp_read(buf, tap) * e;
        tap = sp_wrap(tap + ratio);
        gt++;

        lp += lpCoef * (echo - lp);      /* Tone inside the loop */
        if (lp < SP_DENORM && lp > -SP_DENORM) lp = 0.0f;

        /* soft-clip the feedback: a compounding shift must not run away */
        buf[wp & SP_MASK] = sp_soft(dry + lp * fb);
        wp++;

        float out = dry + (sp_soft(dry + lp) - dry) * mix;
        fxBuf[i]     = out;
        fxBuf[i + 8] = out;
    }

    st->writePos = wp;
    st->tap = tap;
    st->grainT = gt;
    st->lp = sp_flush(lp);
    st->envFast = sp_flush(envF);
    st->envSlow = sp_flush(envS);
    /* Snap the ramp to exactly 1 once it is close enough to hear no difference.
     * Leaving it to converge asymptotically means (1 - ramp) spends the rest of
     * the patch's life as a denormal. */
    st->ramp = (ramp > 0.99999f) ? 1.0f : ramp;
    st->cool = cool;
}

File: tools/audio_preview/renderers/spiral.py
"""Desktop reference renderer for Spiral — a delay whose repeats climb (or sink).

A pitch shift lives INSIDE the feedback loop, so it compounds: repeat 1 is one
step away, repeat 2 is two steps, repeat 3 is three, an endless staircase that
fades out as feedback decays. Shepard-tone territory, but with a small interval
you hear it creep rather than jump.

The tap is grain-re-triggered and read at `ratio`, so one reader both delays and
pitch-shifts (the same trick as Hydra's heads). Two controls sit on that ratio
and cover the two ways to read "the delay rises in pitch":

  Rise   a constant ratio != 1. Discrete: each repeat is one step from the last.
         Bipolar, unity at centre — left of centre the repeats sink instead.
  Glide  depth of an onset-triggered ramp ON that ratio, and Span is how long it
         takes. A note resets the ramp to ZERO and it then climbs toward Glide's
         depth over Span, so the repeats begin at pitch and drift away over a
         while -- a riser. Deliberately not a swoop that settles back.

Set Glide to 0 for a pure staircase; set Rise to centre and Glide up for a pure
glide with no net transposition; use both for a climb that also swoops.

The ratio mapping is piecewise LINEAR, not 2^(semitones/12): the pedal port
cannot call exp2 without pulling in a helper, and equal temperament is not the
point here.

Knobs (manifest order, 0..100):
  Time    delay time, ~40 ms .. 1.2 s
  Feedbk  repeats
  Rise    pitch step per repeat: 0 = octave down, 50 = unity, 100 = octave up
  Glide   how far the pitch travels over the rise (depth)
  Span    how long that rise takes, ~0.2 s .. 8 s
  Tone    low-pass inside the loop, so the climb dulls instead of turning harsh
  Mix     dry/wet
"""

from __future__ import annotations

import numpy as np

GRAIN_MS = 55.0          # tap grain length; short enough to track, long enough to be smooth


def _read(buf, pos, mask):
    i0 = int(pos) & mask
    i1 = (i0 + 1) & mask
    frac = pos - np.floor(pos)
    return buf[i0] * (1.0 - frac) + buf[i1] * frac


def _ratio_from_rise(rise: float) -> float:
    """Piecewise linear, unity at centre: 0 -> 0.5 (oct down), 1 -> 2.0 (oct up)."""
    if rise < 0.5:
        return 0.5 + rise                      # 0.5 .. 1.0
    return 1.0 + (rise - 0.5) * 2.0            # 1.0 .. 2.0


def render(audio, sample_rate, params, tail, root):
    sr = sample_rate
    time_n = float(params.get("time", 45.0)) / 100.0
    fb_n   = float(params.get("feedbk", 60.0)) / 100.0
    rise_n = float(params.get("rise", 62.0)) / 100.0
    glide_n= float(params.get("glide", 45.0)) / 100.0
    span_n = float(params.get("span", 45.0)) / 100.0
    tone_n = float(params.get("tone", 50.0)) / 100.0
    mix    = float(params.get("mix", 50.0)) / 100.0

    delay_s = 0.04 + time_n * time_n * 1.16        # ~40 ms .. 1.2 s, squared for feel
    D = max(int(delay_s * sr), 64)
    fb = fb_n * 0.965                              # long tails, still short of runaway
    base_ratio = _ratio_from_rise(rise_n)
    # Glide pushes the ratio further in whichever direction Rise points.
    glide_dir = 1.0 if rise_n >= 0.5 else -1.0
    glide_amt = glide_n * 0.6
    lp_coef = 0.03 + tone_n * 0.5

    G = max(int(GRAIN_MS * 0.001 * sr), 64)        # grain length
    inv_G = 1.0 / G

    x = audio.astype(np.float64)
    if x.ndim == 1:
        x = x[:, None]
    tail_samps = int(tail * sr) if tail else int(3.0 * sr)
    mono = np.concatenate([x.mean(axis=1), np.zeros(tail_samps)])
    n = mono.shape[0]

    size = 1
    while size < (D + G) * 3:
        size <<= 1
    buf = np.zeros(size)
    mask = size - 1

    wet = np.zeros(n)
    wp = 0
    tap = 0.0
    grain_t = G
    lp = 0.0
    env_f = env_s = 0.0
    ramp = 0.0
    # Span: how long the rise takes. A note resets the ramp to 0 and it climbs
    # toward 1, so the repeats START at pitch and drift away over Span -- the
    # "slowly rises for a while" behaviour, not a swoop settling back.
    span_s = 0.2 + span_n * span_n * 7.8           # ~0.2 s .. 8 s
    ramp_k = 1.0 - np.exp(-1.0 / (span_s * sr))
    a_f, a_s = 0.02, 0.0008
    cool = 0

    for i in range(n):
        dry = mono[i]
        ax = abs(dry)
        env_f += a_f * (ax - env_f)
        env_s += a_s * (ax - env_s)
        # onset: fast envelope jumps clear of the slow one
        if cool <= 0 and env_f > env_s * 1.7 + 0.008:
            ramp = 0.0                             # a note restarts the climb from pitch
            cool = int(0.05 * sr)
        cool -= 1
        ramp += ramp_k * (1.0 - ramp)              # ... and it rises from there

        ratio = base_ratio * (1.0 + glide_amt * ramp * glide_dir)
        if ratio < 0.25:
            ratio = 0.25
        elif ratio > 4.0:
            ratio = 4.0

        if grain_t >= G:                           # re-trigger the tap D behind
            grain_t = 0
            tap = float(wp - D)

        if grain_t < G * 0.5:
            env = grain_t * inv_G * 2.0
        else:
            env = (G - grain_t) * inv_G * 2.0
        env = env * env * (3.0 - 2.0 * env) if env < 1.0 else 1.0

        echo = _read(buf, tap, mask) * env
        tap += ratio
        grain_t += 1

        lp += lp_coef * (echo - lp)                # Tone inside the loop
        e = lp
        # soft-clip the feedback so a compounding shift cannot run away
        f = dry + e * fb
        f = np.clip(f, -1.0, 1.0)
        buf[wp & mask] = 1.5 * f - 0.5 * f ** 3
        wp += 1

        wet[i] = e

    dry_pad = np.concatenate([x, np.zeros((tail_samps, x.shape[1]))], axis=0)
    out = np.empty((n, 2))
    for ch in range(2):
        d = dry_pad[:, min(ch, dry_pad.shape[1] - 1)]
        out[:, ch] = d * (1.0 - mix) + np.clip(d + wet, -1.0, 1.0) * mix
    return out
